README.md
# ProjectCF
A neural network with web-interface.  
Cleaned implementation of https://pqhuy98.hopto.org/ProjectCF/ .  

How it works :  
1)  You upload an image.  
2)  PHP saves it and sends its path to the socket server on port 6969.  
3)  Socket server listens on port 6969 and receive all port receive all the path, read all images and feeds them to a CNN, then sends the result back to PHP by that socket.
4)  PHP decodes the message and generate the result page.

Simple, right ?


history/history.txt
09:57:28 04/02/2017 91.158.230.96 4
10:02:47 04/02/2017 91.158.230.96 5
10:03:09 04/02/2017 91.158.230.96 6
10:03:22 04/02/2017 91.158.230.96 7
10:03:39 04/02/2017 91.158.230.96 8
10:04:01 04/02/2017 91.158.230.96 9
10:05:02 04/02/2017 91.158.230.96 10
10:38:04 04/02/2017 91.158.230.96 0
10:38:26 04/02/2017 91.158.230.96 1
10:38:51 04/02/2017 91.158.230.96 2
10:39:06 04/02/2017 91.158.230.96 3
10:40:07 04/02/2017 91.158.230.96 4
10:40:28 04/02/2017 91.158.230.96 5
11:14:18 04/02/2017 91.158.231.6 6
11:30:12 04/02/2017 91.158.231.6 7
11:39:02 04/02/2017 91.158.231.6 8
11:39:29 04/02/2017 91.158.231.6 9
11:39:42 04/02/2017 91.158.231.6 10
11:40:07 04/02/2017 91.158.231.6 11
11:40:26 04/02/2017 91.158.231.6 12
11:40:41 04/02/2017 91.158.231.6 13
11:41:04 04/02/2017 91.158.231.6 14
12:08:08 04/02/2017 91.158.230.96 15
16:44:59 05/02/2017 91.158.230.96 16
16:45:33 05/02/2017 91.158.230.96 17
16:47:28 05/02/2017 91.158.230.96 18
16:48:06 05/02/2017 91.158.230.96 19
16:48:39 05/02/2017 91.158.230.96 20
16:49:01 05/02/2017 91.158.230.96 21
16:50:19 05/02/2017 91.158.230.96 22
16:50:52 05/02/2017 91.158.230.96 23
16:51:49 05/02/2017 91.158.230.96 24
16:56:02 05/02/2017 91.158.230.96 25
16:56:54 05/02/2017 91.158.230.96 26
16:57:25 05/02/2017 91.158.230.96 27
16:59:15 05/02/2017 91.158.230.96 28
17:01:12 05/02/2017 91.158.230.96 29
17:02:37 05/02/2017 91.158.230.96 30
17:03:20 05/02/2017 91.158.230.96 31
17:03:38 05/02/2017 91.158.230.96 32
18:55:10 05/02/2017 91.158.230.96 33
18:56:25 05/02/2017 91.158.230.96 34
18:59:57 05/02/2017 91.158.230.96 35
19:04:36 05/02/2017 91.158.230.96 36
19:06:02 05/02/2017 91.158.230.96 37
19:07:38 05/02/2017 91.158.230.96 37
19:09:17 05/02/2017 91.158.230.96 38
19:11:00 05/02/2017 91.158.230.96 39
19:11:55 05/02/2017 91.158.230.96 40
19:19:03 05/02/2017 91.158.230.96 41
19:20:04 05/02/2017 91.158.230.96 42
19:21:34 05/02/2017 91.158.230.96 43
19:21:47 05/02/2017 91.158.230.96 44
19:22:11 05/02/2017 91.158.230.96 45
19:23:51 05/02/2017 91.158.230.96 46
19:24:28 05/02/2017 91.158.230.96 47
19:24:43 05/02/2017 91.158.230.96 48
19:25:19 05/02/2017 91.158.230.96 49
19:26:26 05/02/2017 91.158.230.96 50
19:27:00 05/02/2017 91.158.230.96 51
05:38:33 06/02/2017 91.158.230.96 52
05:38:46 06/02/2017 91.158.230.96 53
06:09:42 06/02/2017 91.158.230.96 0
06:10:47 06/02/2017 91.158.230.96 1
06:11:03 06/02/2017 91.158.230.96 2
06:11:41 06/02/2017 91.158.230.96 3
06:15:08 06/02/2017 91.158.230.96 4
06:15:52 06/02/2017 91.158.230.96 5
06:16:05 06/02/2017 91.158.230.96 6
06:18:07 06/02/2017 91.158.230.96 7
06:18:17 06/02/2017 91.158.230.96 8
17:46:48 06/02/2017 91.158.230.96 9
18:01:28 06/02/2017 91.158.230.96 10
18:02:12 06/02/2017 91.158.230.96 11
18:21:10 06/02/2017 91.158.230.96 -1
18:23:50 06/02/2017 91.158.230.96 12
18:25:01 06/02/2017 91.158.230.96 12
19:02:07 06/02/2017 91.158.230.96 18
19:02:20 06/02/2017 91.158.230.96 19
19:02:31 06/02/2017 91.158.230.96 20
19:08:00 06/02/2017 91.158.230.96 21
19:08:15 06/02/2017 91.158.230.96 22
19:09:08 06/02/2017 91.158.230.96 23
19:09:27 06/02/2017 91.158.230.96 24
19:09:36 06/02/2017 91.158.230.96 25
19:09:40 06/02/2017 91.158.230.96 26
19:09:44 06/02/2017 91.158.230.96 27
19:09:47 06/02/2017 91.158.230.96 28
10:45:01 10/02/2017 91.158.230.96 0
10:46:35 10/02/2017 91.158.230.96 1
10:56:03 10/02/2017 91.158.230.96 2
10:56:38 10/02/2017 91.158.230.96 3
03:13:49 13/04/2017 91.158.229.204 33
03:13:57 13/04/2017 91.158.229.204 35
03:13:57 13/04/2017 91.158.229.204 35
03:14:07 13/04/2017 91.158.229.204 37
03:14:13 13/04/2017 91.158.229.204 39
03:14:20 13/04/2017 91.158.229.204 41
03:52:48 17/04/2017 91.158.229.204 43
03:53:42 17/04/2017 91.158.229.204 45
03:53:50 17/04/2017 91.158.229.204 47
03:54:00 17/04/2017 91.158.229.204 49
03:55:48 17/04/2017 91.158.229.204 51
03:55:48 17/04/2017 91.158.229.204 52
03:55:48 17/04/2017 91.158.229.204 54
03:55:55 17/04/2017 91.158.229.204 57
03:56:02 17/04/2017 91.158.229.204 59
03:56:08 17/04/2017 91.158.229.204 61
04:01:35 17/04/2017 91.158.229.204 63
04:03:38 17/04/2017 91.158.229.204 65
04:06:02 17/04/2017 91.158.229.204 67
04:06:13 17/04/2017 91.158.229.204 69
04:06:45 17/04/2017 91.158.229.204 71
04:08:49 17/04/2017 91.158.229.204 73
04:08:52 17/04/2017 91.158.229.204 75
04:09:03 17/04/2017 91.158.229.204 77
04:09:17 17/04/2017 91.158.229.204 79
04:09:21 17/04/2017 91.158.229.204 81
04:09:51 17/04/2017 91.158.229.204 83
04:10:18 17/04/2017 91.158.229.204 85
04:10:56 17/04/2017 91.158.229.204 87
04:11:39 17/04/2017 91.158.229.204 89
04:13:28 17/04/2017 91.158.229.204 91
04:26:06 17/04/2017 91.158.229.204 93
04:26:18 17/04/2017 91.158.229.204 95
04:26:22 17/04/2017 91.158.229.204 97
04:26:32 17/04/2017 91.158.229.204 99
04:26:38 17/04/2017 91.158.229.204 101
04:26:43 17/04/2017 91.158.229.204 103
04:26:50 17/04/2017 91.158.229.204 105
06:20:04 20/05/2017 91.158.229.204 107
06:33:07 20/05/2017 91.158.229.204 0
06:33:29 20/05/2017 91.158.229.204 2
06:33:48 20/05/2017 91.158.229.204 4
06:34:29 20/05/2017 91.158.229.204 6
06:35:38 20/05/2017 91.158.229.204 8
06:35:43 20/05/2017 91.158.229.204 10
06:38:35 20/05/2017 91.158.229.204 12
06:38:42 20/05/2017 91.158.229.204 14
09:24:03 21/05/2017 80.223.84.185 16
09:24:49 21/05/2017 80.223.84.185 18
09:25:23 21/05/2017 80.223.84.185 20
09:28:29 21/05/2017 80.223.84.185 22
10:07:04 21/05/2017 80.223.84.185 24


js/javascript.js
var spinning = false;
var a = 0;

function start_spin() {
	show_helper();
	spinning = true;
	one_step();
}
function stop_spin() {
	hide_helper();
	spinning = false;
}

function one_step() {
/* 	if (spinning==false) return;
	a+=Math.random()*10;
	$("#helper").rotate({angle:a});
	setTimeout(function() {one_step()},10); */
}
function show_helper() {
	$("#helper").rotate({angle:0});
	$("#hp1").attr("class","helper_on");
	$("#hp2").attr("class","helper_on");
}
function hide_helper() {
	$("#helper").rotate({angle:180});
	$("#hp1").attr("class","helper_off");
	$("#hp2").attr("class","helper_off");
}

var idx = 0;
var dur = 2000;
var talk = [
	["I'm thinking ! Please wait...<br>", 3000],
	["Too long ? Let me tell you a joke.<br>", 1500],
]

var jokes = [
	"Man and God met somewhere;<br>Both exclaimed, \"My creator!\"",

	"When I lost my rifle, the Army charged me $85.<br>\
	That’s why in the Navy, the captain goes down with the ship.",

	"Every Scooby-Doo episode would literally be two minutes long\
	<br>if the gang went to the mask store first and asked a few questions.",

	"Red sky at night, shepherd’s delight.<br>Blue sky at night, day.",

	"The closest a person ever comes to perfection<br>\
	is when he fills out a job application form.",

	"I hate Russian dolls,<br>they're so full of themselves.",
	"My wife and I were happy for 20 years,<br>then we met.",
	"Whiteboards are remarkable.<br>",

	"The old lady at an ATM asked me to help check her balance.<br>\
	So I pushed her over.",

	"Best two words joke ever: your life.<br>",

	"I bet you I could stop gambling.<br>",

	"There are two rules for success:<br>1) Don't tell all you know.",
	
	"Why was six scared of seven?<br>\
	Because seven \"ate\" nine.",
]

function next_joke() {
	if (jokes.length == 0) return "Did you see my Javascript ?";
	else {
		var idx = randint(0,jokes.length-1);
		var res = jokes[idx];
		jokes.splice(idx, 1);
		return res;
	}
}

function talking() {
	$("#talk").css("color", "#F4FF81");
	$("#talk").css("letter-spacing", "2px");
	if (idx<talk.length) {
		$("#talk").html(talk[idx][0]);
		setTimeout(function(){talking()}, talk[idx][1]);
		idx++;
	} else {
		$("#talk").css("cursor","pointer");
		$("#talk").html(next_joke());
		// $("#talk").on("click", function() {
		// 	x = next_joke();
		// 	$("#talk").html(x);
		// });
	}
}

$("#helper").rotate({angle:180});
var no_more = false;
$("#test").on("click",function() {
	if (no_more) return;
	no_more = true;
	$("#helper").attr("src","loading.gif");
	startEffect(20, 0);
	talking();
});

setTimeout(function(){startEffect(12, 3000)}, 3000);
$("#form").submit(function() {
	if (no_more) return;
	no_more = true;
	$("#helper").attr("src","loading.gif");
	talking();
});
//Helper functions
function gebi(x) {return document.getElementById(x);}
function randint(min,max) {
    return Math.floor(Math.random()*(max-min+1)+min);
}

socket_server.py
import numpy as np
import matplotlib.pyplot as plt
from keras.models import load_model
from scipy.misc import imread,imresize
import socket
import threading
from PIL import Image

imagepath = "image.jpeg";

def show(im,name=""):
    plt.figure()
    plt.title(name)
    plt.imshow(im)
    plt.show()
    
def toRGB(im, bg_colour=(255, 255, 255)):
    if im.mode in ('RGBA', 'LA') or (im.mode == 'P' and 'transparency' in im.info):
	print "toRGB !"
        alpha = im.convert('RGBA').split()[-1]
        bg = Image.new("RGBA", im.size, bg_colour + (255,))
        bg.paste(im, mask=alpha)
        return bg

    else:
        return im

def crop_and_resize(input_path,height=32,width=32) :
    img = Image.open(input_path)

    width,height = img.size

    mx = max(width,height)
    mn = min(width,height)
    left = (mx-height)/2
    top = (mx-width)/2
    right = width-(left+mn)
    bottom = height-(top+mn)

    img = img.crop((left,top,width-right,height-bottom))

    img = img.convert("RGB")
    img = toRGB(img) 

    img.thumbnail((32,32),Image.ANTIALIAS)
    #show(img)
    
    width, height = img.size
    img = np.reshape(img.getdata(),(32,32,3))
    return img

labels = np.loadtxt('labels.txt', str, delimiter='\n');
print "Loading model"
model = load_model("net.h5")

def get_result(path) :
    print "Rescaling image",path
    try :
        img = crop_and_resize(path)
        img = np.transpose(img,axes=(2,0,1)).astype("float32")
        img/=255
        res = ""
        print "Feed forward..."
        pred = model.predict(np.array([img]))
        print labels[np.argmax(pred)]
        for i in range(pred.shape[1]) :
            x = pred[0,i]
            res+=str(round(x,3))+" "
        return res[:-1]
    except :
        return "-1"

def solve(sock) :
    path = sock.recv(1000000)
    msg = get_result(path)
    sock.send(msg)
    sock.close()
    print msg
    print "Closed"
    print 

print "Creating socket..."
sock = socket.socket()
sock.bind(("",2222))
sock.listen(100)
print "Waiting for connection..."
try :
    while True :
        (client_sock,address) = sock.accept()
        print "Connected by %s"%str(address)
        solve(client_sock)
except KeyboardInterrupt:
    print "Connection closed."
    sock.close()
sock.close()


classify.php
<?php
	function get_ip() {
		$ipaddress = '';
		if (getenv('HTTP_CLIENT_IP'))
			$ipaddress = getenv('HTTP_CLIENT_IP');
		else if(getenv('HTTP_X_FORWARDED_FOR'))
			$ipaddress = getenv('HTTP_X_FORWARDED_FOR');
		else if(getenv('HTTP_X_FORWARDED'))
			$ipaddress = getenv('HTTP_X_FORWARDED');
		else if(getenv('HTTP_FORWARDED_FOR'))
			$ipaddress = getenv('HTTP_FORWARDED_FOR');
		else if(getenv('HTTP_FORWARDED'))
		   $ipaddress = getenv('HTTP_FORWARDED');
		else if(getenv('REMOTE_ADDR'))
			$ipaddress = getenv('REMOTE_ADDR');
		else
			$ipaddress = 'UNKNOWN';
		return $ipaddress;
	}
	function get_result($dir,$path) {
		$real_path=$dir.$path;
 		if (($sock = socket_create(AF_INET,SOCK_STREAM,SOL_TCP))===False) {
			echo "</table><h1>Server error 001</h1>";
			exit;
		}
		if (socket_connect($sock,"192.168.0.169",2222)===False) {
			echo "</table><h1>Service is under maintenance.</h1>";
			exit;
		}
		if (socket_write($sock,$real_path,strlen($real_path))===False) {
			echo "</table><h1>Server error 002</h1>";
			exit;
		}
		$result = socket_read($sock,1000000);
		socket_close($sock);
		if ($result===False) {
			echo "</table><h1>Server doesn't respond.</h1>";
			exit;
		}
		return $result;
	}
	function get_sorted_prob($path,$dir) {
		if(!file_exists("history/"))
			mkdir("history", 0711);
		file_put_contents("history/history.txt",date("H:i:s d/m/Y")." ".get_ip()." ".$path."\n",FILE_APPEND);
		if(!file_exists($dir))
			mkdir($dir, 0777);
		$prob = explode(" ",get_result($dir,$path));
		if ($prob[0]=="-1") {
			return False;
		}
		$res = Array();
		$label = array("Airplane","Car","Bird","Cat","Deer","Dog","Frog","Horse","Ship","Truck");
		for($i=0;$i<10;$i++)
			$res[] = Array("label"=>$label[$i],"prob"=>round(floatval($prob[$i])*150));
		usort($res,function($i,$j) {return $i["prob"]>$j["prob"];});
		$res = array_reverse($res);
		$res[0]["label"] = strtoupper($res[0]["label"]);
		return Array($res[0],$res[1],$res[2],$res[3],$res[4]);
		// $res = Array();
		// $res[] = Array("label"=>"Car","prob"=>75.0);
		// $res[] = Array("label"=>"Dog","prob"=>35.0);
		// $res[] = Array("label"=>"Cat","prob"=>15.0);
		// $res[] = Array("label"=>"Frog","prob"=>5.0);
		// $res[] = Array("label"=>"Ship","prob"=>1.0);
		// return $res;
	}
	function alert($s) {
		echo"<script>window.onload=function(){alert('$s');window.location = 'index.html';}</script>";
	}
	function print_result($res) {
			echo "<table align=center>";
			$b = "<b>";$_b="</b>";
			foreach($res as $p) {
				echo "<tr><td>$b".$p["label"]."$_b</td><td style='text-align:left'>";
				echo "<img src='bar.png' width=".$p["prob"]."px height=10px>";
				echo "</td></tr>";
				$b="";$_b="";
			}
			echo "</table>";
	}
	function upload($file) {
		$upload_dir="raw_images/";
		if(!file_exists($upload_dir))
			mkdir($upload_dir, 0777);
		$name = $file["name"];
		if($file["error"]<>0) {
			switch($file["error"]) {
				case 4:
					return false;
				case 1: 
					alert("<b>$name</b> exceeds the max upload size");
					break;
				case 2:
					alert("<b>$name</b> exceeds the max upload size");
					break;
				case 3:
					alert("<b>$name</b> wasn't uploaded completely.");
					break; 
				case 6: 
					alert("Missing a temporary folder.");
					break;
				case 7:
					alert("System error : failed to write <b>$name</b> to disk.");
					break;
				case 8:
					alert("System error : failed to write <b>$name</b> to disk.");
					break;
			}
		} else if (!exif_imagetype($file["tmp_name"]))
			alert("This image is not valid or corrupted.");
		else {
			$idx = max(0,count(scandir($upload_dir))-2);
			move_uploaded_file($file["tmp_name"],$upload_dir.$idx);
			file_put_contents($upload_dir.$idx.".png","");
			chmod($upload_dir.$idx, 0777);
			chmod($upload_dir.$idx.".png", 0777);
			$res = get_sorted_prob($idx,$upload_dir);
			if ($res!==False) {
				echo "<img src=$upload_dir$idx height=200px class='result_img' onmouseenter='' onmouseleave=''><br><br>";
				print_result($res);
			} else {	
				echo "</table><h1>The image is corrupted.</h1>";
			}
		}
		return true;
	}
	function upload_url($url) {
		$upload_dir="raw_images/";
		if (empty($url)) return false;
		if (getimagesize($url)) {
			if (exif_imagetype($url)==IMAGETYPE_GIF) {
				echo "</table><h1>GIF images are not supported !</h1>";
				return true;
			}
			if(!file_exists($upload_dir))
				mkdir($upload_dir, 0777);
			$idx = max(0,count(scandir($upload_dir))-2);
			copy($url,$upload_dir.$idx);
			file_put_contents($upload_dir.$idx.".png","");
            chmod($upload_dir.$idx,0777);
            chmod($upload_dir.$idx.".png",0777);
			$res = get_sorted_prob($idx,$upload_dir);
			if ($res!==False) {
				echo "<img src=$upload_dir$idx height=200px class='result_img' onmouseenter='' onmouseleave=''><br><br>";
				print_result($res);
			} else {
				echo "</table><h1>The image is corrupted.</h1>";
			}
			return true;
		} else return false;
	}
	function solve() {		
		$msg1 = false;
		echo "<table><tr><td width=500px>";
		if (isset($_FILES["file"]))
			$msg1 = upload($_FILES["file"]);
		$msg2 = false;
		echo "</td><td width=50%>";
		if (isset($_POST["url"]))
			$msg2 = upload_url($_POST["url"]);
		echo "</td></tr></table>";
		if (!$msg1 && !$msg2)
			echo "<img src='oops.jpg' class='result_img'";
		$_POST = Array();
	}
?>


<!DOCTYPE HTML>
<html>
<head>
	<title>VERSION 4.00</title>
	<link rel="icon" href="../cifar10.png">
	<meta charset="utf-8">
	<link rel="stylesheet" href="../css/style.css">
	<link rel="stylesheet" href="css/style.css">
	<link rel="stylesheet" href="../css/projects.css">
	<link href="https://fonts.googleapis.com/css?family=Montserrat|Open+Sans|Pangolin|Roboto|Bahiana|Caveat+Brush|Orbitron" rel="stylesheet">
	<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
	<script src="js/rotate.js"></script>
</head> 
<body>
<div class="go_to_home">
	<a href="../" title="Home"><img src="../home.png" width=25></a>
</div>
<div class="middle">
<div class="welcome_cifar">
<!-- ######################################################################################################### -->
	<span class="header_cifar"><b>"VERSION 4.00"</b></span><br>
	<span class="credit_cifar"><i>I'm a convolutional neural net.<br>Trust me, I can think.</i></span>
	<br><br><br>
	<table align=center width=675px><tr>
	<td width=40%>
s		<!--<div class="helper_off" id="hp1">
			You give me an <b>image</b>.<br>
			I tell you what it is.<br>
			It must be 1 in <b>10</b> classes.<br>
			Accuracy 87% in lastest test.<br>
			<b>If I don't know the answer,<br>
			I'll throw a random result.<br></b>
		</div>-->
	</td><td width=10% style="vertical-align: top">
		<!--<img src="symbol.png" width=100px class="helper_symbol" id="helper" onmouseenter="start_spin()" onmouseleave="stop_spin()">-->
		<?php solve();?>
	</td><td width=40%>
		<!--<div class="helper_off" id="hp2">
			<b><i>Available classes :</i></b><br>
			airplane, bird,<br>
			car, cat,<br>
			deer, dog,<br>
			frog, horse<br>
			ship, truck.
		</div>-->
	</td></tr></table>

	<script src="js/javascript.js"></script>
	<br><br>
	<a href="./" class="back_cifar"><b>Want to test me more ?</b></a>
<!-- ######################################################################################################### -->
</div></div>
<div class="network">
	<div id="large-header" class="large-header"></div>
	<canvas  id="canvas"></canvas>
	<script src="../background/js/TweenLite.min.js"></script>
	<script src="../background/js/EasePack.min.js"></script>
	<script src="../background/js/background.js"></script>
</div>
</body>
</html>

